README.md
# Nations Tournament Tool

This is a tool for helping manage the MabiWeb Nations "[Small Tournament](https://5f99237e5765c.site123.me/)".


### Usage

This tool expects tournament game names in the following syntax:  `[prefix] - Group [A-Z] - Game [xxx]`.

For example, `7.Small Tournament - Group A - Game 105`.

Before running the tool, all games for the tournament should already be created. Each game should have 3 players.

The tool has two tabs:  

#### Get Game List Tab

![Game List Tab](images/game_list_tab.png)

Enter the prefix to generate a file containing the list of games to track for the tournament.

This can only find games that were active within the last 30 days since MabiWeb deletes inactive games after that point.


#### Game Info Tab

![Game Info Tab](images/game_info_tab.png)

Provide a file containing the list of games to track for the tournament.  This can either be the file generated in the `Get Game List` tab, or the one generated by this tab.

The scoring mode determines how tournament scores are calculated.  In "Normal" mode, unfinished games are scored like finished games.  In "Final" mode, unfinished games are scored as unfinished matches per the [tournament's rules](https://5f99237e5765c.site123.me/).

Click the `Download Game Info` tab to generate a file containing all the data on the current games in the tournament.


CHANGELOG.md
## [2.2.1](https://github.com/markwoon/NationsTournamentTool/compare/v2.2.0...v2.2.1) (2023-06-21)


### Bug Fixes

* add time frame field when sending invites ([696767b](https://github.com/markwoon/NationsTournamentTool/commit/696767bbfd3dc4df3f515c50f3cad7cadb637f2f))

## [2.2.0](https://github.com/markwoon/NationsTournamentTool/compare/v2.1.3...v2.2.0) (2023-06-19)


### Features

* support new-format 3P games ([e90c6ab](https://github.com/markwoon/NationsTournamentTool/commit/e90c6ab6fbfd08e20860deca67fab9e507c0a4da))

### [2.1.3](https://github.com/markwoon/NationsTournamentTool/compare/v2.1.2...v2.1.3) (2022-03-09)


### Bug Fixes

* specify TLS 1.2 ([d08c3cf](https://github.com/markwoon/NationsTournamentTool/commit/d08c3cf1050905db149429d4bf88e8fab9960fd6))

### [2.1.2](https://github.com/markwoon/NationsTournamentTool/compare/v2.1.1...v2.1.2) (2022-03-07)


### Bug Fixes

* fix invite button text ([ff53869](https://github.com/markwoon/NationsTournamentTool/commit/ff538698f1edceec6e8e3492c3c4d0b6d5b10ee4))

### [2.1.1](https://github.com/markwoon/NationsTournamentTool/compare/v2.1.0...v2.1.1) (2022-03-03)


### Bug Fixes

* support dynamic # of players ([0c79ac9](https://github.com/markwoon/NationsTournamentTool/commit/0c79ac974fee464900985c63763e24076eec4524))

## [2.1.0](https://github.com/markwoon/NationsTournamentTool/compare/v2.0.2...v2.1.0) (2022-03-01)


### Features

* add support for sending invitiations ([6ab255d](https://github.com/markwoon/NationsTournamentTool/commit/6ab255d8225960e630d5cbad8a759de410e38684))


### Bug Fixes

* fix info message ([d0580bc](https://github.com/markwoon/NationsTournamentTool/commit/d0580bca8abc0ca405bcee3c05072424727db82e))
* remove test prefix ([15b2196](https://github.com/markwoon/NationsTournamentTool/commit/15b21961d0debc832d4fdcba0c19f57e26ac2358))

### [2.0.2](https://github.com/markwoon/NationsTournamentTool/compare/v2.0.1...v2.0.2) (2022-02-26)


### Bug Fixes

* create test 4p games for now ([e36b76c](https://github.com/markwoon/NationsTournamentTool/commit/e36b76cd254be8becec89807e588c89e531a170e))
* fix 4P game creation ([cce9bee](https://github.com/markwoon/NationsTournamentTool/commit/cce9bee0087f91cc4c83472cc693d8124e4ca5fd))

### [2.0.1](https://github.com/markwoon/NationsTournamentTool/compare/v2.0.0...v2.0.1) (2022-02-25)


### Bug Fixes

* support specifying level when creating 4P games ([a98d129](https://github.com/markwoon/NationsTournamentTool/commit/a98d1294b73799036e82cb25a308f8f448ff74e3))

## [2.0.0](https://github.com/markwoon/NationsTournamentTool/compare/v1.4.1...v2.0.0) (2022-02-25)


### ⚠ BREAKING CHANGES

* require Java 17

### Features

* support creating 4p games ([e859430](https://github.com/markwoon/NationsTournamentTool/commit/e85943083d10adac0d9d28f469764918e8bda67a))


### Miscellaneous Chores

* upgrade to Java 17 ([2149af9](https://github.com/markwoon/NationsTournamentTool/commit/2149af9bf44b106f8129e6ab4faa08dd228d1189))

### [1.4.1](https://github.com/markwoon/NationsTournamentTool/compare/v1.4.0...v1.4.1) (2021-08-05)


### Bug Fixes

* fix game group pattern ([2d7c19a](https://github.com/markwoon/NationsTournamentTool/commit/2d7c19ac6da568ea4e9b8a3062ad90d5d12af468))

## [1.4.0](https://github.com/markwoon/NationsTournamentTool/compare/v1.3.6...v1.4.0) (2021-07-22)


### Features

* switch to creating games depending on number of players ([bc8de59](https://github.com/markwoon/NationsTournamentTool/commit/bc8de59297f5639a8fa020681be31dc5c597eb31))

### [1.3.6](https://github.com/markwoon/NationsTournamentTool/compare/v1.3.5...v1.3.6) (2021-06-01)


### Bug Fixes

* improve setup phase handling ([be9df3d](https://github.com/markwoon/NationsTournamentTool/commit/be9df3df6e5ec350a572c45f71ba5b40b0e3c6bb))

### [1.3.5](https://github.com/markwoon/NationsTournamentTool/compare/v1.3.4...v1.3.5) (2021-06-01)


### Bug Fixes

* handle setup phase ([0615746](https://github.com/markwoon/NationsTournamentTool/commit/061574621d6b4916cbe649105a877fe3f5777164))

### [1.3.4](https://github.com/markwoon/NationsTournamentTool/compare/v1.3.3...v1.3.4) (2021-05-12)


### Bug Fixes

* improve support for different date formats and reading tsv that has been exported from Excel ([8883135](https://github.com/markwoon/NationsTournamentTool/commit/88831355c8d8498e4642bf1eea93c05a86079cfe))

### [1.3.3](https://github.com/markwoon/NationsTournamentTool/compare/v1.3.2...v1.3.3) (2021-05-10)


### Bug Fixes

* fix TSV parsing when not all expected columns are present ([3763ba3](https://github.com/markwoon/NationsTournamentTool/commit/3763ba3cc00377786470a0fd3bbcff84ab29a283))

### [1.3.2](https://github.com/markwoon/NationsTournamentTool/compare/v1.3.1...v1.3.2) (2021-05-04)


### Bug Fixes

* avoid NPE when getting list ([886e79d](https://github.com/markwoon/NationsTournamentTool/commit/886e79def48161ca464c9ec8495970f2f4a0ea28))

### [1.3.1](https://github.com/markwoon/NationsTournamentTool/compare/v1.3.0...v1.3.1) (2021-04-28)


### Bug Fixes

* fix build by switching from logback to slf4j-nop ([75e5f93](https://github.com/markwoon/NationsTournamentTool/commit/75e5f9380cdc55a90ccc74d7af439386fd53c419))

## [1.3.0](https://github.com/markwoon/NationsTournamentTool/compare/v1.2.4...v1.3.0) (2021-04-28)


### Features

* add support for creating games from app ([0db17a6](https://github.com/markwoon/NationsTournamentTool/commit/0db17a64d046acb09f20164017edca27d27d8527))
* update exported columns and scoring calc ([1e85f7f](https://github.com/markwoon/NationsTournamentTool/commit/1e85f7fb1647596f52d8f5c22b4553a6d74e5f43))


### Bug Fixes

* handle expired games properly ([cdcba70](https://github.com/markwoon/NationsTournamentTool/commit/cdcba70f2ecc361951625564b12f16d3b2424390))


src/main/java/org/markwoon/nations/MabiWebHelper.java
package org.markwoon.nations;

import java.io.IOException;
import java.net.CookieManager;
import java.net.URI;
import java.net.URLEncoder;
import java.net.http.HttpClient;
import java.net.http.HttpRequest;
import java.net.http.HttpResponse;
import java.nio.charset.StandardCharsets;
import java.time.Duration;
import java.time.LocalDateTime;
import java.time.format.DateTimeFormatter;
import java.util.ArrayList;
import java.util.List;
import java.util.Objects;
import java.util.SortedMap;
import java.util.TreeMap;
import java.util.regex.Matcher;
import java.util.regex.Pattern;
import com.google.common.base.Preconditions;
import com.google.common.base.Splitter;
import com.google.common.net.HttpHeaders;
import org.apache.commons.lang3.StringUtils;
import org.jsoup.Jsoup;
import org.jsoup.nodes.Document;
import org.jsoup.nodes.Element;
import org.jsoup.nodes.Node;
import org.jsoup.select.Elements;
import org.markwoon.nations.model.Game;


/**
 * Utilities for working with MabiWeb.
 *
 * @author Mark Woon
 */
public class MabiWebHelper {
  public enum Level {EMPEROR, KING, PRINCE, CHIEFTAIN}
  private static final Splitter sf_commaSplitter = Splitter.on(",").trimResults()
      .omitEmptyStrings();
  // maximum allowed days is 30
  private static final String sf_finishedGamesUrl =
      "http://www.mabiweb.com/modules.php?name=Game_Manager&op=completed_games&gm_type=10&days=30";
  // maximum allowed days is 30
  private static final String sf_runningGamesUrl =
      "http://www.mabiweb.com/modules.php?name=Game_Manager&op=running_games&gm_type=10&days=30";


  /**
   * Gets the basics of all games for a tournament.
   */
  public static SortedMap<String, Game> getGames(String tournamentName) throws IOException {

    SortedMap<String, Game> games = new TreeMap<>();
    getGames(tournamentName, games, false);
    getGames(tournamentName, games, true);
    return games;
  }

  private static void getGames(String tournamentName, SortedMap<String, Game> games,
      boolean finished) throws IOException {

    String url;
    if (finished) {
      url = sf_finishedGamesUrl;
    } else {
      url = sf_runningGamesUrl;
    }
    Document doc = Jsoup.connect(url)
        .get();
    for (Element tr : doc.select("#gamemanager-gamelists > table > tbody > tr")) {
      Elements tds = tr.select("td");
      String name = tds.get(2).text();
      if (name.startsWith(tournamentName)) {
        String id = tds.get(0).text();
        Game game = new Game(id, name);

        if (finished) {
          game.setRound(Game.ROUND_FINISHED);
          game.setPlayers(sf_commaSplitter.splitToList(tds.get(4).text()));
        } else {
          String data = tds.get(4).text();
          int idx = data.indexOf((" ("));
          game.setPlayers(sf_commaSplitter.splitToList(data.substring(0, idx)));
          game.setRoundString(data.substring(idx + 2, data.length() - 1));
        }

        String lastUpdate = tds.get(5).text();
        game.setLastUpdated(calculateTimestamp(lastUpdate));
        games.put(name, game);
      }
    }
  }


  private static final Pattern sf_timeAgoPattern = Pattern.compile("(\\d+)(?:\\.(\\d+))? (min|hours|days) ago");

  private static LocalDateTime calculateTimestamp(String lastUpdate) throws IOException {
    Matcher m = sf_timeAgoPattern.matcher(lastUpdate);
    if (m.matches()) {
      int main = Integer.parseInt(m.group(1));
      int sub = 0;
      if (m.group(2) != null) {
        sub = Integer.parseInt(m.group(2));
      }
      LocalDateTime dateTime = LocalDateTime.now();
      switch (m.group(3)) {
        case "min" -> {
          dateTime = dateTime.minusMinutes(main);
          if (sub > 0) {
            dateTime = dateTime.minusSeconds((long)6 * sub);
          }
        }
        case "hours" -> {
          dateTime = dateTime.minusHours(main);
          if (sub > 0) {
            dateTime = dateTime.minusMinutes((long)6 * sub);
          }
        }
        case "days" -> {
          dateTime = dateTime.minusDays(main);
          if (sub > 0) {
            dateTime = dateTime.minusMinutes((long)(2.4 * sub * 60));
          }
        }
      }
      return dateTime;

    } else {
      throw new IOException("Unexpected time ago format: " + lastUpdate);
    }
  }


  /**
   * Gets the details of all games.
   */
  public static void getGameDetails(SortedMap<String, Game> games) throws IOException {
    LastUpdatedHelper lastUpdatedHelper = new LastUpdatedHelper();
    for (String key : games.keySet()) {
      Game game = games.get(key);
      boolean gotDetails = getGameDetails(game);
      boolean gotLastUpdated = lastUpdatedHelper.updateLastUpdated(game);
      // if can't get details and can't get last updated, game has expired on MabiWeb
      // mark game as dead if it isn't already finished
      if (!gotDetails && !gotLastUpdated && !game.isFinished()) {
        game.setRound(Game.ROUND_DEAD);
      }
    }
  }


  private static final Pattern sf_roundPattern = Pattern.compile("round: (.*?), Action phase");
  private static final Pattern sf_playerTurnPattern = Pattern.compile("<b>(.*?)</b>");
  private static final Pattern sf_playerFinalScorePattern = Pattern.compile("(\\w+)\\s+(\\d+)\\s+\\(.+?\\)");
  private static final Pattern sf_vpTablePattern =
      Pattern.compile("<TABLE class=\\\\'vp_summary_table\\\\'.*?>(.*)</TABLE>");
  private static final Pattern sf_vpPattern =
      Pattern.compile("<TR><TD.*?><IMG.*?> (.+?)</TD><TD.*?>(\\d+?) <IMG.*?></TD></TR>");
  private static final Pattern sf_countryPattern =
      Pattern.compile("modules/GM_Nations/images/DPlayerBoard_(\\w+)\\.jpg");
  private static final Pattern sf_topPattern = Pattern.compile("top:(\\d+)");
  private static final Pattern sf_startedPattern = Pattern.compile("Nations Game ID=\\d+, started ([\\d.]+, \\d+:\\d+:\\d+)");
  private static final Pattern sf_finishedPattern = Pattern.compile("Game finished ([\\d.]+, \\d+:\\d+:\\d+)");
  private static final Splitter sf_spaceSplitter = Splitter.on(" ").trimResults()
      .omitEmptyStrings();
  private static final DateTimeFormatter sf_gameDetailDateTimeFormatter =
      DateTimeFormatter.ofPattern("yyyy.MM.dd, HH:mm:ss");

  /**
   * Gets the details of a specific game.
   */
  public static boolean getGameDetails(Game game) throws IOException {
    Document doc = Jsoup.connect("http://www.mabiweb.com/modules.php?name=GM_Nations&op=view_game_reset&g_id=" + game.getId())
        .get();

    if (doc.text().contains("Error: can't load position for game with ID")) {
      // game has not started
      return false;
    }

    Element header = Objects.requireNonNull(doc.getElementById("nations-gameheader"));
    String headerPlayerInfo = header.text().substring(header.text().indexOf("Players:") + 8);
    if (headerPlayerInfo.contains("Player Boards selected for this game:")) {
      headerPlayerInfo = headerPlayerInfo.substring(0,
          headerPlayerInfo.indexOf("Player Boards selected for this game:"));
    }
    headerPlayerInfo = StringUtils.stripToEmpty(headerPlayerInfo);
    Element lastActions = Objects.requireNonNull(doc.getElementById("last_actions"));

    boolean isFinished = header.html().contains("Game finished");
    // finished?
    if (isFinished) {
      // round
      game.setRound(Game.ROUND_FINISHED);

      // started
      if (game.getGameStarted() == null) {
        Matcher startedMatcher = sf_startedPattern.matcher(lastActions.text());
        if (startedMatcher.find()) {
          game.setGameStarted(LocalDateTime.parse(startedMatcher.group(1), sf_gameDetailDateTimeFormatter));
        }
      }
      // finished
      Matcher finishedMatcher = sf_finishedPattern.matcher(lastActions.text());
      if (finishedMatcher.find()) {
        game.setGameFinished(LocalDateTime.parse(finishedMatcher.group(1), sf_gameDetailDateTimeFormatter));
      }

      // players and score
      Matcher m = sf_playerFinalScorePattern.matcher(headerPlayerInfo);
      List<String[]> playerData = new ArrayList<>();
      while (m.find()) {
        playerData.add(new String[] { m.group(1), m.group(2) });
      }
      List<String> players = new ArrayList<>();
      for (String[] p : playerData) {
        players.add(p[0]);
        game.setVp(p[0], Integer.parseInt(p[1]));
      }
      game.setPlayers(players);
      game.setActivePlayer(null);

    } else {
      // round
      Matcher roundMatcher = sf_roundPattern.matcher(header.text());
      if (roundMatcher.find()) {
        game.setRoundString(roundMatcher.group(1));
      }

      // started
      Matcher startedMatcher = sf_startedPattern.matcher(lastActions.text());
      if (startedMatcher.find()) {
        game.setGameStarted(LocalDateTime.parse(startedMatcher.group(1), sf_gameDetailDateTimeFormatter));
      }

      // players
      game.setPlayers(sf_spaceSplitter.splitToList(headerPlayerInfo));
      // active player
      String players = header.html().substring(header.html().indexOf("Players:"));
      Matcher playerMatcher = sf_playerTurnPattern.matcher(players);
      if (playerMatcher.find()) {
        game.setActivePlayer(playerMatcher.group(1));
      }

      if (!game.isSettingUp()) {
        // score
        Matcher vpTableMatcher = sf_vpTablePattern.matcher(doc.html());
        if (vpTableMatcher.find()) {
          String vpTable = vpTableMatcher.group(1);
          Matcher vpMatcher = sf_vpPattern.matcher(vpTable);
          while (vpMatcher.find()) {
            game.setVp(vpMatcher.group(1), Integer.parseInt(vpMatcher.group(2)));
          }
        }
      }
    }

    if (!game.isSettingUp()) {
      // country
      for (String player : game.getPlayers()) {
        Element board = Objects.requireNonNull(doc.getElementById(player));
        Matcher countryMatcher = sf_countryPattern.matcher(board.html());
        if (countryMatcher.find()) {
          game.setCountry(player, countryMatcher.group(1));
        }

        if (!isFinished) {
          Elements elems = board.select("div.outline_text");
          for (Element elem : elems) {
            Node prev = Objects.requireNonNull(elem.previousSibling());
            if (prev.attr("src").contains("/images/Meeple_")) {
              Matcher topMatcher = sf_topPattern.matcher(elem.attr("style"));
              if (topMatcher.find()) {
                if (Integer.parseInt(topMatcher.group(1)) > 400) {
                  game.setUnusedWorkers(player, Integer.parseInt(elem.text()));
                }
              }
            }
          }
        }
      }
    }
    return true;
  }


  /**
   * Helper class to get the last updated timestamp for a game.
   * Caches finished/running game lists.
   */
  private static class LastUpdatedHelper {
    private final Document m_finishedGames;
    private final Document m_runningGames;

    LastUpdatedHelper() throws IOException {
      m_finishedGames = Jsoup.connect(sf_finishedGamesUrl)
          .get();
      m_runningGames = Jsoup.connect(sf_runningGamesUrl)
          .get();
    }

    boolean updateLastUpdated(Game game) throws IOException {
      Document doc = game.isFinished() ? m_finishedGames : m_runningGames;
      for (Element tr : doc.select("#gamemanager-gamelists > table > tbody > tr")) {
        Elements tds = tr.select("td");
        String name = tds.get(2).text();
        if (name.startsWith(game.getName())) {
          String lastUpdate = tds.get(5).text();
          game.setLastUpdated(calculateTimestamp(lastUpdate));
          return true;
        }
      }
      return false;
    }
  }


  private final HttpClient m_httpClient;

  public MabiWebHelper() {
    m_httpClient = HttpClient.newBuilder()
        .cookieHandler(new CookieManager())
        .build();
  }

  public boolean login(String userId, String pwd) throws IOException, InterruptedException {
    HttpRequest request = HttpRequest.newBuilder()
        .uri(URI.create("http://www.mabiweb.com/modules.php?name=Your_Account"))
        .timeout(Duration.ofSeconds(30))
        .header(HttpHeaders.CONTENT_TYPE, "application/x-www-form-urlencoded")
        .POST(HttpRequest.BodyPublishers.ofString("op=login" +
            "&username=" + URLEncoder.encode(userId, StandardCharsets.UTF_8) +
            "&user_password=" + URLEncoder.encode(pwd, StandardCharsets.UTF_8)))
        .build();
    HttpResponse<String> response = m_httpClient.send(request, HttpResponse.BodyHandlers.ofString());
    if (response.statusCode() != 302) {
      return false;
    }
    return isLoggedIn();
  }

  public boolean isLoggedIn() throws IOException, InterruptedException {
    HttpRequest request = HttpRequest.newBuilder()
        .uri(URI.create("http://www.mabiweb.com/modules.php?name=Your_Account&stop=1"))
        .timeout(Duration.ofSeconds(30))
        .build();
    HttpResponse<String> response = m_httpClient.send(request, HttpResponse.BodyHandlers.ofString());
    return !response.body().contains("<b>Login Incorrect! Please Try Again...</b>");
  }


  public boolean createGame(String name, String password, int numPlayers, Level level)
      throws IOException, InterruptedException {

    int levelNum;
    switch (level) {
      case CHIEFTAIN -> levelNum = 4;
      case PRINCE -> levelNum = 3;
      case KING -> levelNum = 2;
      case EMPEROR -> levelNum = 1;
      default -> throw new IllegalArgumentException("Unsupported level: " + level);
    }

    HttpRequest request = HttpRequest.newBuilder()
        .uri(URI.create("http://www.mabiweb.com/modules.php?name=GM_Nations"))
        .timeout(Duration.ofSeconds(30))
        .header(HttpHeaders.CONTENT_TYPE, "application/x-www-form-urlencoded")
        .POST(HttpRequest.BodyPublishers.ofString("op=new_game" +
            "&num_allowed[]=" + numPlayers +
            "&title=" + URLEncoder.encode(name, StandardCharsets.UTF_8) +
            "&password=" + URLEncoder.encode(password, StandardCharsets.UTF_8) +
            "&level=" + levelNum +
            // players pick boards
            "&boards=2" +
            // advanced cards
            "&cards[0]=1" +
            // expert cards
            "&cards[1]=1" +
            // dynasties expansion
            "&dynasties=1"
        ))
        .build();
    HttpResponse<String> response = m_httpClient.send(request, HttpResponse.BodyHandlers.ofString());
    String body = response.body();
    System.out.println(body);
    return body.contains("Game created");
  }



  public static List<NewGame> buildTournamentGroup3pGameList(String prefix, String group,
      int numPlayers) {
    prefix = StringUtils.stripToNull(prefix);
    Preconditions.checkNotNull(prefix);
    group = StringUtils.stripToNull(group);
    Preconditions.checkNotNull(group);
    Preconditions.checkArgument(group.length() == 1);

    int subgroups;
    int gamesPerSubgroup;
    switch (numPlayers) {
      case 9 -> {
        subgroups = 1;
        gamesPerSubgroup = 12;
      }
      case 12 -> {
        subgroups = 1;
        gamesPerSubgroup = 16;
      }
      case 15 -> {
        subgroups = 1;
        gamesPerSubgroup = 20;
      }
      case 18 -> {
        subgroups = 2;
        gamesPerSubgroup = 12;
      }
      case 21 -> {
        subgroups = 1;
        gamesPerSubgroup = 28;
      }
      case 24 -> {
        subgroups = 2;
        gamesPerSubgroup = 16;
      }
      case 27 -> {
        subgroups = 3;
        gamesPerSubgroup = 12;
      }
      default -> throw new UnsupportedOperationException("Cannot have a group with " + numPlayers +
          " players");
    }

    if (subgroups == 1) {
      return buildTournamentGameList(prefix, group, 0, gamesPerSubgroup);
    }
    List<NewGame> games = new ArrayList<>();
    for (int x = 0; x < subgroups; x += 1) {
      games.addAll(buildTournamentGameList(prefix, group, x + 1, gamesPerSubgroup));
    }
    return games;
  }

  public static List<NewGame> buildTournamentGroup3pV2GameList(String name, Level level,
      int period, int numMatches) {
    Preconditions.checkNotNull(name);
    Preconditions.checkArgument(period > 0 && period < 10);
    Preconditions.checkArgument(numMatches > 0 && numMatches < 100);

    List<NewGame> games = new ArrayList<>();
    for (int gameNum = 0; gameNum < numMatches; gameNum += 1) {
      String gameId = period + String.format("%02d", gameNum + 1);
      games.add(new NewGame(name + " - " + gameId, gameId, level));
    }
    return games;
  }

  public static List<NewGame> buildTournamentGroup4pGameList(String prefix, int division,
      String group, Level level) {
    prefix = StringUtils.stripToNull(prefix);
    Preconditions.checkNotNull(prefix);
    group = StringUtils.stripToNull(group);
    Preconditions.checkNotNull(group);
    Preconditions.checkArgument(group.length() == 1);

    int groupNum = group.toUpperCase().charAt(0) - 'A' + 1;

    List<NewGame> games = new ArrayList<>();
    for (int gameNum = (groupNum - 1) * 20 + 1, numGames = 0; numGames < 20;
         gameNum += 1, numGames += 1) {
      String gameId = division + String.format("%02d", gameNum);
      games.add(new NewGame(prefix + " - Game " + gameId, gameId, level));
    }
    return games;
  }



  public static List<NewGame> buildTournamentGameList(String prefix, String group, int subgroup,
      int numGames) {
    prefix = StringUtils.stripToNull(prefix);
    Preconditions.checkNotNull(prefix);
    group = StringUtils.stripToNull(group);
    Preconditions.checkNotNull(group);
    Preconditions.checkArgument(group.length() == 1);
    Preconditions.checkArgument(subgroup > -1 && subgroup < 9);

    List<NewGame> games = new ArrayList<>();
    String groupName = prefix + " - Group " + group.toUpperCase();
    int groupNum = group.toUpperCase().charAt(0) - 'A' + 1;
    int subMod = 0;
    if (subgroup > 0) {
      groupName += subgroup;
      subMod = (subgroup - 1) * numGames;
    }
    int gamesPerLevel = numGames / 4;
    for (int gameNum = 1; gameNum <= numGames; gameNum += 1) {
      Level level = Level.CHIEFTAIN;
      if (gameNum <= gamesPerLevel) {
        level = Level.EMPEROR;
      } else if (gameNum <= 2*gamesPerLevel) {
        level = Level.KING;
      } else if (gameNum <= 3*gamesPerLevel) {
        level = Level.PRINCE;
      }
      String gameId = groupNum + String.format("%02d", gameNum + subMod);
      games.add(new NewGame(groupName + " - Game " + gameId, gameId, level));
    }
    return games;
  }

  static final class NewGame {
    String name;
    String password;
    Level level;

    NewGame(String name, String password, Level level) {
      this.name = name;
      this.password = password;
      this.level = level;
    }

    @Override
    public String toString() {
      return name + "    (password='" + password + "', level=" + level + ")";
    }
  }
}
